File: LeetCode_Java/array/CarPooling.java
package array;

public class CarPooling {

    public boolean carPooling(int[][] trips, int capacity) {
        int[] timestamp = new int[1001];

        for (int[] trip: trips) {
            int passengers = trip[0];
            int start = trip[1];
            int end = trip[2];

            timestamp[start] += passengers;
            timestamp[end] -= passengers;
        }

        int totalPassengers = 0;
        for (int p: timestamp) {
            totalPassengers += p;
            if (totalPassengers > capacity) {
                return false;
            }
        }

        return true;
    }

    public static void main(String[] args) {
        CarPooling solution = new CarPooling();

        // Test case: trips = [[2,1,5],[3,3,7]], capacity = 4
        int[][] trips1 = {{2, 1, 5}, {3, 3, 7}};
        int capacity1 = 4;
        // Expected output: false
        System.out.println(solution.carPooling(trips1, capacity1));

        // Test case: trips = [[2,1,5],[3,3,7]], capacity = 5
        int[][] trips2 = {{2, 1, 5}, {3, 3, 7}};
        int capacity2 = 5;
        // Expected output: true
        System.out.println(solution.carPooling(trips2, capacity2));
    }
}

File: LeetCode_Python/prefixsum/CarPooling.py
class CarPooling:
    def carPooling(self, trips: list[list[int]], capacity: int) -> bool:
        events = []

        for passengers, start, end in trips:
            events.append((start, passengers))
            events.append((end, -passengers))

        events.sort()

        current_capacity = 0

        for _, passengers in events:
            current_capacity += passengers
            if current_capacity > capacity:
                return False

        return True


# Time Complexity: O(nlogn)
# Space Complexity: O(n)

if __name__ == "__main__":
    solution = CarPooling()

    print("Expected Output : False")
    print("Actual Output : ", solution.carPooling([[2, 1, 5], [3, 3, 7]], 4))

    print("Expected Output : True")
    print("Actual Output : ", solution.carPooling([[2, 1, 5], [3, 3, 7]], 5))

    print("Expected Output : True")
    print("Actual Output : ", solution.carPooling([[2, 1, 5], [3, 5, 7]], 3))
